File: file/workspace/p230127/src/InstanceEx01.java
/*
 * 	[클래스] class
 * 	- 자료형 중 하나
 *	 	
 * 	기본 자료형(primitive type) : int, float, double, ...
 * 	사용자 정의 자료형(참조 자료형,reference type) : String, class, ...
 * 	
 * 	- 클래스명은 대문자 시작(반드시 문자로 시작)
 * 	- 클래스 내 멤버 : 변수, 메소드, 클래스
 *	
 *	[객체] object, instance
 * 	- 클래스로 생성된 구체화(실체화)된 대상
 * 	- Java는 모든 객체를 heap 영역에 생성 -> 멤버도 heap 영역에 할당
 * 	- 객체는 heap 영역에 생성되면, 할당된 주소를 리턴(반환)
 * 	- 생성된 객체는 반드시 주소를 통해 접근(참조 변수에 저장된 주소)
 * 	- Java는 멤버 변수를 자동으로 초기화(주소값은 null로 저장)	
 * 
 * 	- 객체 생성 문법
 * 	new(heap영역에 할당시키는 연산자) 클래스명();
 * 
 * 	- 참조(reference) 변수 선언 문법
 * 	클래스명 변수명;	
 * 
 */


public class InstanceEx01 {
	public static void main(String[] args) {
		Student lee = new Student(); //생성된 객체는 참조변수에 저장된 주소값을 통해 접근
		lee./*접근 연산자*/age=27;
		lee.name="Jung Hyeon";
		lee.printInfo();
		lee.study();
		
		Student leeTwo = lee;
		System.out.println("[leeTwo] age >> "+leeTwo.age);
	}
}

// [클래스 정의]
class Student{
	int age;
	String name;
	void study() {
		System.out.println("공부 중 입니다...");
	}
	void printInfo() {
		System.out.println("이름 : "+name);
		System.out.println("나이 : "+age);
	}
}

File: python/p230419/instance_ex01.py
class Student:
    # def __init__(self):
    #     print('매개변수 1개인 생성자 실행됨!')
    
    # def __init__(self, name, age):
    #     print('생성자 실행됨!')
    #     self.name = name
    #     self.age = age

    def __init__(self, *args):
        print('생성자 실행됨')
        self.name, self.age = args

    def greeting(self):
        print('안녕하세요.')

    def setName(self, name):
        self.name = name

    def showInfo(self):
        print('제 이름은 ' + self.name + '입니다.')
        print(f'나이는 {self.age}세입니다.')

# 메소드 호출시 인수로 참조변수에 저장된 주소가 첫 번째로 무조건 넘어감
# 메소드에서 인스턴스 변수를 사용하기 위해 주소를 전달
print('[hong]')
hong = Student('홍길동', 29)
hong.greeting()     
# hong.setName('홍길동')
hong.showInfo()

print('\n[park]')
park = Student('박보검', 35)
park.greeting()
# park.setName('박보검')
park.showInfo()

print('\n[lee]')
# lee = Student()